sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/dao/BigtableChangeStreamAccessor.java

```diff
@@ -0,0 +1,220 @@
+      bigtableAccessors = new ConcurrentHashMap<>();
+
+  private static final ConcurrentHashMap<BigtableConfig, AtomicInteger> refcounts =
+      new ConcurrentHashMap<>();
+
+  private final BigtableDataClient dataClient;
+  private final BigtableTableAdminClient tableAdminClient;
+  private final BigtableInstanceAdminClient instanceAdminClient;
+  private final BigtableConfig bigtableConfig;
+
+  private BigtableChangeStreamAccessor(
+      BigtableDataClient dataClient,
+      BigtableTableAdminClient tableAdminClient,
+      BigtableInstanceAdminClient instanceAdminClient,
+      BigtableConfig bigtableConfig) {
+    this.dataClient = dataClient;
+    this.tableAdminClient = tableAdminClient;
+    this.instanceAdminClient = instanceAdminClient;
+    this.bigtableConfig = bigtableConfig;
+  }
+
+  /**
+   * Create a BigtableAccess if it doesn't exist and store it in the cache for faster access. If it
+   * does exist, just return it.
+   *
+   * @param bigtableConfig config that contains all the parameters to connect to a Cloud Bigtable
+   *     instance
+   * @return data and admin clients connected to the Cloud Bigtable instance
+   * @throws IOException if the connection fails
+   */
+  public static synchronized BigtableChangeStreamAccessor getOrCreate(
+      @NonNull BigtableConfig bigtableConfig) throws IOException {
+    if (bigtableAccessors.get(bigtableConfig) == null) {
+      BigtableChangeStreamAccessor bigtableAccessor =
+          BigtableChangeStreamAccessor.createAccessor(bigtableConfig);
+      bigtableAccessors.put(bigtableConfig, bigtableAccessor);
+      refcounts.putIfAbsent(bigtableConfig, new AtomicInteger(0));
+    }
+    checkStateNotNull(refcounts.get(bigtableConfig)).incrementAndGet();
+    return checkStateNotNull(bigtableAccessors.get(bigtableConfig));
+  }
+
+  private static BigtableChangeStreamAccessor createAccessor(@NonNull BigtableConfig bigtableConfig)
+      throws IOException {
+    String projectId = checkArgumentNotNull(bigtableConfig.getProjectId()).get();
+    String instanceId = checkArgumentNotNull(bigtableConfig.getInstanceId()).get();
+    String appProfileId = checkArgumentNotNull(bigtableConfig.getAppProfileId()).get();
+    BigtableDataSettings.Builder dataSettingsBuilder = BigtableDataSettings.newBuilder();
+    BigtableTableAdminSettings.Builder tableAdminSettingsBuilder =
+        BigtableTableAdminSettings.newBuilder();
+    BigtableInstanceAdminSettings.Builder instanceAdminSettingsBuilder =
+        BigtableInstanceAdminSettings.newBuilder();
+
+    dataSettingsBuilder.setProjectId(projectId);
+    tableAdminSettingsBuilder.setProjectId(projectId);
+    instanceAdminSettingsBuilder.setProjectId(projectId);
+
+    dataSettingsBuilder.setInstanceId(instanceId);
+    tableAdminSettingsBuilder.setInstanceId(instanceId);
+
+    if (appProfileId != null) {
+      dataSettingsBuilder.setAppProfileId(appProfileId);
+    }
+
+    RetrySettings.Builder readRowRetrySettings =
```

Comment:
why are we override retry settings of all these operations?

Reply:
Because the default don't work well for us if we are checkpointing frequently. A connection might hang or fall in the outliers and is very slow to respond. We want to recover quickly from that.